File: Chromatico/CCOTextLayer.m
//
//  CCOTextLayer.m
//  Chromatico
//
//  Created by Justin Sheckler on 4/13/14.
//  Copyright (c) 2014 Justin Sheckler. All rights reserved.
//

#import <QuartzCore/QuartzCore.h>

#import "CCOTextLayer.h"
#import "CCOConstants.h"
#import "CCOPalette.h"
#import "CATextLayer+AutoSizing.h"

@interface CCOTextLayer ()

@property CATextLayer *textLayer;

@end

@implementation CCOTextLayer

- (void)createLayer
{
    self.textLayer = [[CATextLayer alloc] init];
    self.layer = self.textLayer;
}

- (void)setUpGeometry
{
    CGFloat outerWidth = self.outerBounds.size.width;
    CGFloat outerHeight = self.outerBounds.size.height;
    
    [CATransaction begin];
    [CATransaction setAnimationDuration:0.0];
    CGFloat fontSize = (CGFloat)arc4random_uniform(outerHeight) + (outerHeight / 10);
    CGFloat zPosition = (CGFloat)arc4random_uniform(100) + 100;
    CGColorRef color = [self.palette grabColor];
    NSString *fontName = [CCOConstants getRandomFont];

    self.textLayer.string = [CCOConstants getRandomString];
    self.textLayer.font = (CFStringRef)CFBridgingRetain(fontName);
    self.textLayer.fontSize = fontSize;
    self.textLayer.foregroundColor = color;
    self.textLayer.zPosition = zPosition;
    
    // sets layer origin to middle point
    [self.textLayer adjustBoundsToFit];
    
    self.xPos = outerWidth + (self.layer.bounds.size.width / 2);
    self.yPos = (float)arc4random_uniform(outerHeight);
    self.xFinal = 0 - (self.layer.bounds.size.width / 2);

    self.textLayer.position = CGPointMake(self.xPos, self.yPos);
    [CATransaction commit];
}

- (void)changeColor
{
    CGColorRef color = [self.palette grabColor];
    float duration = [CCOConstants getColorAnimationDuration];
    
    [CATransaction begin];
    [CATransaction setAnimationDuration:duration];
    [CATransaction setAnimationTimingFunction:[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear]];
    self.textLayer.foregroundColor = color;
    [CATransaction commit];
    
}

@end

File: Chromatico/CCOTextLayer.swift
//
//  CCOTextLayer.swift
//  Chromatico
//
//  Created by Justin Sheckler on 6/14/14.
//  Copyright (c) 2014 Justin Sheckler. All rights reserved.
//

import Foundation
import QuartzCore

class CCOTextLayer: CCOLayer {
    
    var textLayer: CATextLayer!
    
    init(palette: CCOPalette!) {
        textLayer = CATextLayer()
        super.init(palette: palette, layer: textLayer)
    }
    
    override func setUpGeometry() {
        CATransaction.begin()
        CATransaction.setAnimationDuration(CFTimeInterval(0.0))
        
        let outerWidth = CGFloat(CGRectGetWidth(outerBounds))
        let outerHeight = CGFloat(CGRectGetHeight(outerBounds))
        
        let fontSize = CGFloat(CCORandom.rand(Int(outerHeight)) + (Int(outerHeight) / 10))
        
        let color = palette.grabColor()
        let zPosition = CGFloat(CCORandom.rand(100) + 100)

        textLayer.string = CCOConstants.getRandomString()
        textLayer.font = CCOConstants.getRandomFont()
        textLayer.fontSize = fontSize
        textLayer.foregroundColor = color
        textLayer.zPosition = zPosition
        
        // sets layer origin to middle point
        textLayer.adjustBoundsToFit()
        let innerWidth = CGFloat(CGRectGetWidth(self.layer.bounds))
        
        xPos = outerWidth + innerWidth / 2
        yPos = CGFloat(CCORandom.rand(Int(outerHeight)))
        xFinal = 0 - (innerWidth / 2)
        
        textLayer.position = CGPointMake(xPos, yPos)
        
        CATransaction.commit()
    }
    
}